applications/convection_diffusion_application/custom_conditions/thermal_face.cpp

```diff
@@ -0,0 +1,430 @@
+    const double aux_conv_rhs = rData.ConvectionCoefficient * (gauss_pt_unknown - rData.AmbientTemperature);
+    for (unsigned int i = 0; i < (this->GetGeometry()).PointsNumber(); ++i) {
+        // Add external face heat flux contribution
+        rRightHandSideVector[i] += rData.N(i) * gauss_pt_flux * rData.Weight;
+        // Add ambient radiation contribution
+        rRightHandSideVector[i] -= rData.N(i) * aux_rad_rhs * rData.Weight;
+        // Add ambient temperature convection contribution
+        rRightHandSideVector[i] -= rData.N(i) * aux_conv_rhs * rData.Weight;
+    }
+}
+
+void ThermalFace::AddIntegrationPointLHSContribution(
+    MatrixType& rLeftHandSideMatrix,
+    const ConditionDataStruct &rData)
+{
+    const double stefen_boltzmann = 5.67e-8;
+    const unsigned int n_nodes = (this->GetGeometry()).PointsNumber();
+    const double aux_rad_lhs = rData.Emissivity * stefen_boltzmann * 4.0 * std::pow(rData.GaussPointUnknown(), 3);
+    for (unsigned int i = 0; i < n_nodes; ++i) {
+        for (unsigned int j = 0; j < n_nodes; ++j) {
+            // Add ambient radiation contribution
+            rLeftHandSideMatrix(i,j) += rData.N(i) * aux_rad_lhs * rData.N(j) * rData.Weight;
+            // Ambient temperature convection contribution
+            rLeftHandSideMatrix(i,j) += rData.N(i) * rData.ConvectionCoefficient * rData.N(j) * rData.Weight;
+        }
+    }
+}
+
+void ThermalFace::FillConditionDataStructure(
+    const ProcessInfo &rCurrentProcessInfo,
+    ConditionDataStruct &rData)
+{
+    // Set user-defined variables values in data container
+    ConvectionDiffusionSettings &r_conv_diff_settings = *(rCurrentProcessInfo[CONVECTION_DIFFUSION_SETTINGS]);
+    const Variable<double> &r_unknown_var = r_conv_diff_settings.GetUnknownVariable();
+    const Variable<double> &r_flux_var = r_conv_diff_settings.GetSurfaceSourceVariable();
+    const auto &r_geometry = this->GetGeometry();
+    const unsigned int n_nodes = r_geometry.PointsNumber();
+    rData.UnknownValues.resize(n_nodes, false);
+    rData.FaceHeatFluxValues.resize(n_nodes, false);
+
+    for (unsigned int i = 0; i < n_nodes; ++i) {
+        rData.UnknownValues[i] = r_geometry[i].FastGetSolutionStepValue(r_unknown_var);
+        rData.FaceHeatFluxValues[i] = r_geometry[i].FastGetSolutionStepValue(r_flux_var);
+    }
+
+    // Check (and resize) and fill data container arrays
+    if (rData.UnknownValues.size() != n_nodes) {
+        rData.UnknownValues.resize(n_nodes, false);
+    }
+    if (rData.FaceHeatFluxValues.size() != n_nodes) {
+        rData.FaceHeatFluxValues.resize(n_nodes, false);
+    }
+
+    // Fill data container values from properties
+    // Const reference is required to have thread-safe access
+    const auto &r_prop = this->GetProperties();
+    rData.Emissivity = r_prop[EMISSIVITY];
+    rData.AmbientTemperature = r_prop[AMBIENT_TEMPERATURE];
+    rData.ConvectionCoefficient = r_prop[CONVECTION_COEFFICIENT];
+}
+
+// template <>
```

Comment:
I would suggest to remove unused code, particularly in this case, where the normal can be computed by the geometry

Reply:
Agree! I forgot to remove it...

Reply:
Done. Now the normal is computed using the geometry. Thanks for noticing.